musicApp/R/functions.R

```diff
@@ -167,3 +177,115 @@ getAlbumLyrics = function(albumname, artist, progressBar = FALSE)
+
+makePlaylist <- function(playlistID){
+
+  # create dataframe with data about the playlist and create variable with list of artists.
+  playlist = get_playlist_audio_features('Karel Veldkamp', playlistID)
+  artists = playlist %>%
+    unnest(track.artists) %>%
+    select(name, id) %>%
+    unique
+
+
+  # sample 30 artists (if n > 30)
+  n = nrow(artists)
+  if (n > 30)
+  {
+    sample <- sample(1:n, 30)
+  }else
+  {
+    sample <- 1:n
+  }
+
+  # create a pool of artist that have collaborated with the artists in the playlist, or that are related to them.
+  artistPool = data.frame()
+  for (i in sample)
+  {
+    collabs = get_artist_albums(artists[i, 'id']) %>%
+      select(artists) %>%
+      unnest(artists) %>%
+      filter(!name %in% artists$name) %>%
+      select(id) %>%
+      unique()
+
+    related = get_related_artists(artists[i, 'id'])[1:5,] %>%
+      filter(!name %in% artists$name) %>%
+      select(id)
+
+    artistPool = rbind(artistPool, collabs, related)
+  }
+
+  # get top tracks from each artist in the artist pool
+  allTracks = data.frame()
+  for (i in 1:nrow(artistPool))
+  {
+    toptracks <- get_artist_top_tracks(artistPool[i,])
+    if (length(toptracks)>0){
+      if (! toptracks$id %in% allTracks$id){
+        tracks <- get_track_audio_features(toptracks$id)
+        allTracks <- rbind(allTracks, tracks)
+      }
+    }
+  }
+
+  # get matrix with scaled audio features.
+  audioFeatures = as.matrix(allTracks[features])
+  audioFeatures = scale(audioFeatures)
+  rownames(audioFeatures) <- allTracks$uri
+
+  # to pick which songs are most similar to the ones in the user playlist. I use k means clustering:
+
+  # first save the audio features of songs in user playlist (scaled to a z score)
+  features = c('danceability', 'energy', 'loudness', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence')
+  playlistProfile = scale(playlist[features])
+
+
+  # then calculate the mean silhouette coefficient for differnt numbers of k to decide the optimal k value:
+  scores = 0  # first score is set to 0 because sillouette coefficient is not defined for k == 1
+  maxCluster = min(nrow(playlistProfile)-1,10) # try up to 10 means
+
+  # this loop tries k means clustering with 2 up to 10 clusters:
+  for (k in 2:maxCluster)
+  {
+    kmeans = kmeans(playlistProfile,k)
+    sCoefs = silhouette(kmeans$cluster, dist(playlistProfile))[,3]
+    scores = c(scores, mean(sCoefs))
+  }
+
+  # I then see which k lead to the highest silhouette score and use that model.
+  k = which.max(scores)
+  model = kmeans(playlistProfile, k)
+
+  # determine how much songs should be chosen for each cluster:
+  clusterSize <- count(model$cluster)$freq / length(model$cluster)
+  clusterSongs <- floor(clusterSize * 30)
+
+  # finally, i loop through the amount of clusters to see how scose each song is to each cluster
+  uris <- c()
+  for (i in 1:k)
+  {
+    # calculate euclidean distance of each song to the cluster center:
```

Comment:
Rather trivial but shortly justify why you used the euclidean distance as the relevant metric instead of Manhatten distance for instance?